### DISCUSSION TITLE
How does ObjectLock work on Filesystem Level?

### DISCUSSION THREAD

**Discussion by tsohst** [minio/minio#18493] [Q&A] [answered]
Can someone tell me how ObjectLock works on filesystem level using xfs?
I know that while using s3 commands, ObjectLock works as intended and your unable to delete objects.

But what if an attacker gain access to the filesystem?
Is an "rm" or "dd" or any other linux command able to break object lock?
Obviously, if you have access to the server, you can perform something like `rm -rf / --no-preserve-root` to destroy the root filesystem.
But if I have minio on seperate disks and seperate mountpaths and all buckets are protected with object lock - would it be possible to recover minio?

**Comment by klauspost** [comment]
@tsohst I don't understand your question. How would you via "seperate disks and seperate mountpaths" prevent deletes, if you assume the attacker has root access?

Your best protection is server-side replication. At least then the attacker would need to have root access to both clusters.

**Reply by tsohst** [reply]
Okay, so I was wondering, how secure object lock or WORM in general really is on the filesystem layer?
If I understand you correct, data is just immutable with the S3 protocol? 
If I can access to the filesystem layer I can erase the data anyway?

**Reply by klauspost** [reply, marked as answer]
@tsohst MinIO does not (and should not) have root access, so we have to abide by user-space rules.

MinIO protects against mutations through the APIs. It does not have any "super-root" capabilities.

I don't think you'll find any software that can protect your data if your box has been rooted.
